File: Course_1-Algorithmic_Toolbox/Week-6/Excercise_Challenges/2_partitioning_souvenirs/partition3.py
import numpy as np

def partition3(A):
    sums = sum(A)
    W = sums // 3

    if W * 3 != sums:
        return 0

    n = len(A)
    dp = np.zeros((n + 1, W + 1))

    for i in range(n + 1):
        for j in range(W + 1):
            if A[i - 1] <= j:
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - A[i - 1]] + A[i - 1])
            else:
                dp[i][j] = dp[i - 1][j]

    return W == int(dp[n][W])

if __name__ == '__main__':
    n = int(input())
    A = list(map(int, input().split()))
    
    result = partition3(A)
    print(1 if result else 0)

File: Course_1-Algorithmic_Toolbox/Week-6/Excercise_Challenges/2_partitioning_souvenirs/partition3.cpp
#include <iostream>
#include <vector>
using namespace std;

bool partition3(int total, vector<int> &A)
{
    int W = total / 3;
    if(W * 3 != total)
        return 0;
    
    int n = A.size();
    vector<vector<int> > dp(n + 1, vector<int>(W+1));

    for(int i = 0; i < n + 1; ++i){
        for(int j = 0; j < W + 1; ++j){
            if(i == 0 || j == 0)
                dp[i][j] = 0;
            else if(A[i - 1] <= j)
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - A[i - 1]] + A[i - 1]);
            else
                dp[i][j] = dp[i - 1][j];
        }
    }

    return (dp[n][W] == W);
}

int main(void)
{
    int n;
    cin >> n;
    vector<int> A(n);
    int total = 0;

    for(int i = 0; i < n; ++i){
        cin >> A[i];
        total += A[i];
    }
    cout << partition3(total, A) << '\n';

    return 0;
}
